Implement user sign up

tfidfwastaken opened the pull request:
Closes [sc-110]: As a Guest, I want to sign up for an account, so I can create and share forms

This finishes the user sign up stories. Unfortunately, the diff is a bit bloated because of all the boilerplate needed to set up the database, specs and testing. Since all of those are pretty interconnected, I couldn't find a simple way to split it into smaller, easy to review chunks. Let me know if a synchronous review can speed things up.

The meat of the functionality can be found in 6e3224f8173390856fb044b49823a59b465bad3d.

tfidfwastaken committed:
resources: move css and favicon to public directory

We do this to restrict access to the static resources we serve. With the
previous setup we are giving unrestricted access to our whole resource
path, which is not desirable.

tfidfwastaken committed:
setup reproducible development database

Create a docker-compose build that sets up a development database in
PostgreSQL.

Co-authored-by: Udit Kumar <udit@nilenso.com>

tfidfwastaken committed:
config: setup basic configuration

We use aero to setup configuration for our application.

Co-authored-by: Udit Kumar <udit@nilenso.com>

tfidfwastaken committed:
migrations: create user account table

We setup migrations using migratus and create a user account table for
our application.

Along with that we also provide a command line interface for performing
basic migration tasks.

Co-authored-by: Udit Kumar <udit@nilenso.com>

tfidfwastaken committed:
docs: add instructions for local development

tfidfwastaken committed:
views: add sign up page

The sign up page will send a POST request to the /signup route.

tfidfwastaken committed:
Implement sign ups

This commit sets up all the boilerplate and scaffolding required to
implement sign up as well.

Passwords are stored using the argon2id algorithm of the `buddy-hashers`
library.

We create middleware to wrap exceptions and keywordize form parameters.

We also write specs that are used for validation of form and user data.

We use Ring's flash middleware to inform the user if the sign up failed.

tfidfwastaken committed:
signup: test signup helpers and custom middleware

tfidfwastaken committed:
Add docstrings to public functions

ssrihari commented on src/bugle_forms/config.clj:
```diff
@@ -0,0 +1,16 @@
+   [aero.core :refer [read-config]]
+   [clojure.java.io :as io]))
+
+(defn profile
```
Consider renaming to env/environment?

ssrihari commented on src/bugle_forms/config.clj:
```diff
@@ -0,0 +1,16 @@
+(ns bugle-forms.config
+  (:require
+   [aero.core :refer [read-config]]
```
Prefer aliases wherever possible.

ssrihari commented on src/bugle_forms/core.clj:
```diff
@@ -1,29 +1,37 @@
 (ns bugle-forms.core
```
[nit] "Core" might not be a good name. For anything.

ssrihari commented on src/bugle_forms/core.clj:
```diff
@@ -1,29 +1,37 @@
   (reset! server
           (raj/run-jetty app
-                         {:port (Integer. (or (System/getenv "PORT") "8080"))
+                         {:port (:port (config))
```
1. Read config once, story in a read-only state.
2. Look for a simpler API like `(config/lookup :port)`

ssrihari commented on src/bugle_forms/db/connection.clj:
```diff
@@ -0,0 +1,9 @@
+(def db-spec (:db-spec (config)))
+
+(def datasource
+  (jdbc/get-datasource db-spec))
```
Remember these are loaded at compile time. Consider using mount to manage component state.

ssrihari commented on src/bugle_forms/db/utils.clj:
```diff
@@ -0,0 +1,14 @@
+  (not-empty
+   (sql/find-by-keys ds table {key val}
+                     {:table-fn csk/->snake_case
+                      :column-fn csk/->snake_case})))
```
1. Move `table-fn`, `column-fn` out of this function, and perhaps into a config somewhere.
2. `user/find-by-email` might be more domain specific, and appropriate at this time of design.
3. `sql/find-by-keys` is already what you're looking to build.

ssrihari commented on src/bugle_forms/handlers/user.clj:
```diff
@@ -1,9 +1,48 @@
+  [{:keys [flash]}]
+  (response/response
+   (html5 (layout/application {:title "Sign up" :flash flash}
+                              user-views/signup))))
```
View and handler code in same ns?

ssrihari replied:
`(response/response view/sign-up-page)`

ssrihari commented on src/bugle_forms/handlers/user.clj:
```diff
@@ -1,9 +1,48 @@
+        (sql/insert! db/datasource :user-account account
+                     {:table-fn csk/->snake_case})
+        (-> (response/redirect "/login" :see-other)
+            (assoc :flash "Account creation successful!"))))))
```
Read about:
- validation (structural, semantic, domain)
- schemas (plumatic/schema swagger, openAPI, etc)

ssrihari replied:
Move the actual domain logic into its own ns.

ssrihari replied:
Read about MVC.

ssrihari replied:
`user/create`

ssrihari commented on src/bugle_forms/routes.clj:
```diff
@@ -1,11 +1,14 @@
-    "signup" {:get user-handlers/signup}
-    "login" {:get user-handlers/login}
+    "signup" {:get user-handlers/signup
+              :post user-handlers/create-user}
```
```clojure
(def routes
{:post
   {:handler user-handler/create-user
    :schema ""spec/signup-form}
}
})


(def bidi-routes
  (->> routes ...))
```

ssrihari commented on src/bugle_forms/handlers/user.clj:
```diff
@@ -1,9 +1,48 @@
+   [clojure.spec.alpha :as s]
+   [hiccup.page :refer [html5]]
+   [next.jdbc.sql :as sql]
    [ring.util.response :as response]))
```
Use a large list of dependencies as a measure of the concerns in the ns.

ssrihari commented on resources/migrations/20220209094622-create-user.up.sql:
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS user_account (
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
```
consider `users`

ssrihari commented on src/bugle_forms/model/user.clj:
```diff
@@ -0,0 +1,12 @@
+  [{:keys [name email password]}]
+  {:user/name name
+   :user/email email
+   :user/password (hashers/derive password {:alg :argon2id})})
```
`(def ^:private table-name :user-accounts)`

ssrihari commented on src/bugle_forms/views/user.clj:
```diff
@@ -0,0 +1,30 @@
+    [:div {:id "password"}
+     (util/labelled-text-input
+      "Set a password"
+      :required true
```
You might want to derive this information from spec.

ssrihari commented on src/bugle_forms/views/utils.clj:
```diff
@@ -0,0 +1,15 @@
+(defn labelled-text-input
+  "Create a representation of a labelled input box.
+  Takes a label along with an option map which are added as CSS attributes."
+  [label & args]
```
`[label & {:as args}]`

ssrihari replied:
`[label options]`

tfidfwastaken committed:
refactor application

Revisit the separation of concerns as per Srihari's review. The changes
include:

- Add a structural validation layer
- Reorganize the handler to strictly be a controller—move domain
specific actions to user
- Reorganize modules and improve readability
- Renaming 'core' to 'app'

tfidfwastaken committed:
introduce component lifecycle management

We use mount to manage the setup and teardown of global application
states.

tfidfwastaken committed:
tests: add fixtures and test user signup

tfidfwastaken committed:
CI: introduce test database in pipeline

Now that we have tests that connect to a database, we need to setup test
databases in CI as well.

tfidfwastaken committed:
config: simplify configs and abolish env vars

We stop using a checked-in secrets file that reads environment
variables, as that is the worst of both worlds. Let's switch to a
simpler config, that assumes the secret is in a `secrets.edn` file as
before, but we won't read or rely on environment variables to manage
secrets.

This will also mean that our planned docker-compose deployment is able
to support '_FILE' based variables to get access to secrets.

tfidfwastaken committed:
add recipe for docker deployments

Production and staging deployments read a secret config file named
`config.edn` and a password file for provisioning the database named
`db-password.txt` located in the `~/.config/bugle-forms` directory.

tfidfwastaken committed:
CI: remove `pull_request` trigger in test workflow

Since all push events to the repository run tests anyway via staging
workflows, the `pull_request` trigger is redundant. Let's avoid testing
the same code twice.

tfidfwastaken committed:
CI: add workflow to deploy to staging on a VPS

We use Docker Compose to deploy, by issuing a command to the remote
docker daemon using the DOCKER_HOST environment variable.

tfidfwastaken committed:
CI: add workflow to deploy to production on a VPS

We use Docker Compose to deploy, by issuing a command to the remote
docker daemon using the DOCKER_HOST environment variable.

Production deployments only occur on version tag pushes.

kirang89 commented on .github/workflows/deploy-production.yml:
```diff
@@ -0,0 +1,37 @@
+        mkdir -p ~/.ssh/
+        echo "${{ secrets.PROD_PRIVATE_KEY }}" > ~/.ssh/id_rsa
+        sudo chmod 600 ~/.ssh/id_rsa
+        echo "${{ secrets.PROD_KNOWN_HOSTS }}" > ~/.ssh/known_hosts
```
question: What information is configured here to be written into the `known_hosts` file?

tfidfwastaken replied:
The `PROD_KNOWN_HOSTS` secret is grabbed from `ssh-keyscan -t rsa <REMOTE_HOST_NAME>`, which gets the public key of our deployment server in `known_hosts` format.

This lets us verify the authenticity of remote host while bypassing the prompt that usually comes up when connecting to ssh interactively for the first time.

kirang89 commented on README.md:
````diff
@@ -8,17 +8,44 @@ you need to make your analytics, polls or events work.
+lein migrations up <ids ...>     # migrate specified ids
+lein migrations down <ids ...>   # rollback specified ids
+lein migrations create <name>    # generate migration files with specified name
+```
````
suggestion: It'd be a good idea to add a section on Deployment with just enough details for a user to visualise the deployment flow.

kirang89 commented on README.md:
````diff
@@ -8,17 +8,44 @@ you need to make your analytics, polls or events work.
+lein migrations down <ids ...>   # rollback specified ids
+lein migrations create <name>    # generate migration files with specified name
+```
 
````
suggestion: Consider adding a section on Configuration that summarises how configurations are managed and how they could be changed.

tfidfwastaken committed:
signup: improve form validation

Enhancements include:
- Enforce minimum password length
- Allow 1 character tld in emails since they are technically valid.
- Visually confirm valid input on the client side with CSS and HTML
  attributes

kirang89 commented on deploy.yml:
```diff
@@ -0,0 +1,40 @@
+    environment:
+      POSTGRES_USER: bugle_prod
+      POSTGRES_DB: bugle_prod
+      POSTGRES_PASSWORD_FILE: /run/secrets/db_password
```
question: How does this path work to retrieve the db password?

tfidfwastaken replied:
Docker Compose automatically [creates a runtime bind mount](https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose) from the secrets file to the `/run/secrets/<secret-name>` path.

kirang89 commented on deploy.yml:
```diff
@@ -0,0 +1,40 @@
+
+secrets:
+  db_password:
+    file: /home/deploy/.config/bugle-forms/db_password.txt
+  app_secrets:
+    file: /home/deploy/.config/bugle-forms/secrets.edn
```
question: How are these files created?

tfidfwastaken replied:
They are manually copied to the server on initial setup. Like you said, this needs to be documented. Added it as a task for this project.

kirang89 reviewed (commented):
The CI/CD workflow mostly looks good. I've added some minor questions to clarify my understanding of how things work.
